content/posts/building_jnigen.md

```diff
@@ -0,0 +1,262 @@
+
+__TODO: A High level Architecture Diagram:__
+
+## JNI runtime support
+
+JNI is an old native interface [^jni]. Therefore, lot of quirks have to be considered when our goal is to abstract the JNI to a high level language like Dart.
+
+### Differences between platforms
+On Android, Flutter application runs embedded in Android JVM. Therefore the VM already exists. We just initialize the JNI using a plugin, and also obtain a reference to application context.
+
+On standalone targets (Flutter Desktop and standalone Dart), there's no JVM. On initialization, a JVM has to be spawned, using the JRE available on the machine. If you intend to use some external libraries, (like the PDFBox library we use in one of the example), the JAR files must be provided as classpath.
+
+Besides, we have a support library written in C. It's packaged automatically with Flutter apps since it's a native plugin. However, this library's path must be provided with standalone.
+
+99% of real world application of Java interop will be on Android. There are enough quirks in standalone support [^jvm_lazy_load] that one of them can run for the president of a European country. Then why take the pain to implement support for standalone targets? _Having standalone support provides a means to do proper unit testing of support library as well as generated bindings._
+
+### Thread-local this, thread-local that
+JNI is supposed for use by C / C++, not a higher level language like Dart. `JNIEnv` struct, basically a vtable of 200+ functions, provides the entry point for most functionality of the JNI. Unfortunately this struct is valid only in the thread it is obtained.
+
+Dart being a high level language, Thread pinning is [not yet supported](https://github.com/dart-lang/sdk/issues/46943), and we didn't want to rely on implementation details of threading, in presence of async-await in the language. Thus the `JNIEnv *` was wrapped in a `thread_local`. This singleton design was not really a problem, because there can be at most 1 JVM in a process, anyway.
+
+For the same reason, all references returned from JNI are converted to global references before returning to Dart. This works fine in practice. Reference lifecycle is handled by `NativeFinalizer` mechanism, with an API for explicit deletion if required.
+
+### The curious case of the UI thread classloader
+Android has a concept of UI thread. Confusingly, in a Flutter app, this is different from the thread running Flutter application.
+
+Surprisingly, most platform classes are not available if you call JNI's `FindClass` from a flutter thread. This is because those threads have a much barebones classloader. The solution suggested by [Android Developers documentation](https://developer.android.com/training/articles/perf-jni) is to store a reference to original thread's class loader and call its loadClass method instead of JNI's `FindClass`, which worked for us.
```

Comment:
Wait, how did I not know about this? 😄

Reply:
You probably forgot about it. We'd this figured out in initial prototype using `javaparser`. That was last march.